src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,241 @@
+    CHIP_ERROR EncodeSupportedModes(const AttributeValueEncoder::ListEncodeHelper & encoder);
+
+public:
+    /**
+     * Initialise the ModeBase server instance.
+     * @return Returns an error if the cluster ID given is not af a valid ModeBase cluster, if the
+     * given endpoint and cluster ID have not been enabled in zap, if the CommandHandler or
+     * AttributeHandler registration fails or if the AppInit() returns an error.
+     */
+    CHIP_ERROR Init();
+
+    // Attribute setters
+    /**
+     * Sets the Start-Up attribute. Note, this also handles writing the new value into non-volatile storage.
+     * @param aNewStartUpMode The value to which the Start-Up mode is to be set.
+     * @return Returns a ConstraintError if the aNewStartUpMode value is not valid. Returns Success otherwise.
+     */
+    Protocols::InteractionModel::Status UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode);
+
+    /**
+     * Sets the On-Mode attribute. Note, this also handles writing the new value into non-volatile storage.
+     * @param aNewOnMode The value to which the On-Mode mode is to be set.
+     * @return Returns a ConstraintError if the aNewOnMode value is not valid. Returns Success otherwise.
+     */
+    Protocols::InteractionModel::Status UpdateOnMode(DataModel::Nullable<uint8_t> aNewOnMode);
+
+    /**
+     * Sets the Current-Mode attribute. Note, this also handles writing the new value into non-volatile storage.
+     * @param aNewMode The value to which the Current-Mode mode is to be set.
+     * @return Returns a ConstraintError if the aNewMode value is not valid. Returns Success otherwise.
+     */
+    Protocols::InteractionModel::Status UpdateCurrentMode(uint8_t aNewMode);
+
+    // Attribute getters
+    /**
+     * @return The Start-Up mode.
+     */
+    DataModel::Nullable<uint8_t> GetStartUpMode() const;
+
+    /**
+     * @return The On mode.
+     */
+    DataModel::Nullable<uint8_t> GetOnMode() const;
+
+    /**
+     * @return The Current mode.
+     */
+    uint8_t GetCurrentMode() const;
+
+    /**
+     * @return The endpoint ID.
+     */
+    EndpointId GetEndpointId() const { return mEndpointId; }
+
+    // Cluster constants
+    static constexpr uint8_t kMaxModeLabelSize = 64;
+    static constexpr uint8_t kMaxNumOfModeTags = 8;
+
+    /**
+     * Returns true if the feature is supported.
+     * @param feature the feature to check.
+     */
+    bool HasFeature(Feature feature) const;
+
+    /**
+     * This function returns true if the mode value given matches one of the supported modes, otherwise it returns false.
+     * @param mode
+     */
+    bool IsSupportedMode(uint8_t mode);
+
+    /**
+     * Creates a mode base cluster instance. The Init() function needs to be called for this instance to be registered and
+     * called by the interaction model at the appropriate times.
+     * **Note** Once instantiated, the lifetime of this object must span the entire lifetime of the application.
```

Comment:
This note can go away once we add the attribute access interface dereg bit, right?  Please don't lose track of it.

Reply:
Removed it.